relayer/relayer/src/connections/eth.rs

```diff
@@ -1,39 +1,235 @@
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EthVsockSigner {
+    client: Arc<Mutex<signer_client::Client>>,
+    chain_id: u64,
+    address: Address,
+}
+
+impl EthVsockSigner {
+    async fn sign_hash(&self, hash: H256) -> Result<Signature, EthVsockSignerError> {
+        let client = self.client.clone();
+        let signature = spawn_blocking(move || {
+            let client = client.lock().unwrap();
+            client.sign_eth_hash(hash)
+        })
+        .await??;
+        Ok(signature)
+    }
+}
+
+#[async_trait::async_trait]
+impl Signer for EthVsockSigner {
+    type Error = EthVsockSignerError;
+
+    async fn sign_message<S: Send + Sync + AsRef<[u8]>>(
+        &self,
+        message: S,
+    ) -> Result<Signature, Self::Error> {
+        let message = message.as_ref();
+        let message_hash = hash_message(message);
+
+        self.sign_hash(message_hash).await
+    }
+
+    async fn sign_transaction(&self, tx: &TypedTransaction) -> Result<Signature, Self::Error> {
+        let client = self.client.clone();
+        let tx = tx.clone();
+        let signature = spawn_blocking(move || {
+            let client = client.lock().unwrap();
+            client.sign_eth_tx(&tx)
+        })
+        .await??;
+        Ok(signature)
+    }
+
+    async fn sign_typed_data<T: Eip712 + Send + Sync>(
+        &self,
+        payload: &T,
+    ) -> Result<Signature, Self::Error> {
+        let encoded = payload
+            .encode_eip712()
+            .map_err(|e| EthVsockSignerError::Eip712(e.to_string()))?;
+
+        self.sign_hash(H256::from(encoded)).await
+    }
+
+    fn address(&self) -> Address {
+        self.address
+    }
+
+    fn chain_id(&self) -> u64 {
+        self.chain_id
+    }
+
+    fn with_chain_id<T: Into<u64>>(mut self, chain_id: T) -> Self {
+        self.chain_id = chain_id.into();
+        self
+    }
 }
 
 pub async fn connect(url: &str) -> EthConnection {
     Provider::<Http>::connect(url).await
 }
 
-pub async fn sign(
+pub async fn with_local_wallet(
     connection: EthConnection,
     wallet: LocalWallet,
 ) -> Result<SignedEthConnection, EthConnectionError> {
     let nonce_manager = connection.nonce_manager(wallet.address());
     nonce_manager.initialize_nonce(None).await?;
+    let signer = EthereumSigner::Local(wallet);
+
+    Ok(SignerMiddleware::new_with_provider_chain(nonce_manager, signer).await?)
+}
+
+pub async fn with_signer(
+    connection: EthConnection,
+    cid: u32,
+    port: u32,
+) -> Result<SignedEthConnection, EthConnectionError> {
+    let client = signer_client::Client::new(cid, port)?;
+    let client = Arc::new(Mutex::new(client));
+    let client_rc = client.clone();
+    let address = spawn_blocking(move || {
+        let client = client_rc.lock().unwrap();
+        client.eth_address()
+    })
+    .await??;
+
+    let nonce_manager = connection.nonce_manager(address);
```

Comment:
I think you forgot to initialize the nonce manager here

Reply:
:+1: